crates/sparrow-scheduler/src/worker.rs

```diff
@@ -71,15 +76,38 @@ pub struct Worker {
 
 impl Worker {
     /// Run the work loop to completion.
-    pub(crate) fn work_loop(mut self) -> error_stack::Result<(), Error> {
-        while let Some(task) = self.queue.pop() {
-            if task.do_work(&mut self)? {
-                // This means that the task was schedule while we were executing.
-                // As a result, we didn't add it to any queue yet, so we need to
-                // do so now.
-                self.queue.push_global(task);
+    pub(crate) fn work_loop(
+        mut self,
+        index: usize,
+        pending_set: Arc<PendingSet>,
+    ) -> error_stack::Result<(), Error> {
+        loop {
+            while let Some(task) = self.queue.pop() {
+                tracing::info!("Running task: {task:?} on worker {index}");
+                if task.do_work(&mut self)? {
+                    // This means that the task was scheduled while we were executing.
+                    // As a result, we didn't add it to any queue yet, so we need to
+                    // do so now.
+                    debug_println!(
+                        crate::DEBUG_SCHEDULER,
+                        "Task {task:?} scheduled during execution; re-scheduling"
+                    );
+                    self.queue.push_global(task);
+                }
+            }
+
+            let pending_count = pending_set.pending_partition_count();
+            if pending_count == 0 {
+                break;
+            } else {
+                // Right now, this "busy-waits" by immediately trying to pull more work.
```

Comment:
Note this: the busy waiting is definitely not ideal. I may take a quick stab at fixing it in this PR, if not it should be a fast-follow-up.

Reply:
Changed to idle the thread and wake it up when work is queued.